contracts/tests/mcms/ManyChainMultiSigSubgroups.spec.ts

```diff
@@ -0,0 +1,528 @@
+import {
+  toNano,
+  beginCell,
+  Address,
+  Dictionary,
+  TransactionDescriptionGeneric,
+  TransactionComputeVm,
+} from '@ton/core'
+import { Blockchain, SandboxContract, TreasuryContract } from '@ton/sandbox'
+import '@ton/test-utils'
+import { compile } from '@ton/blueprint'
+import { MCMSBaseTestSetup, MCMSTestCode, TestSigner } from './ManyChainMultiSigBaseTest'
+import { merkleProof } from '../../src/mcms'
+import * as mcms from '../../wrappers/mcms/MCMS'
+import { asSnakeData, uint8ArrayToBigInt } from '../../src/utils'
+import { KeyPair, keyPairFromSeed, sha256, sign } from '@ton/crypto'
+import { randomBytes } from 'crypto'
+import { ocr } from '../../wrappers/libraries/ocr'
+import { generateEd25519KeyPair } from '../libraries/ocr/Helpers'
+
+describe('MCMS - ManyChainMultiSigSubgroupsTest', () => {
+  let blockchain: Blockchain
+  let code: MCMSTestCode
+  let acc: {
+    deployer: SandboxContract<TreasuryContract>
+    multisigOwner: SandboxContract<TreasuryContract>
+    signers: SandboxContract<TreasuryContract>[]
+  }
+  let bind: {
+    mcms: SandboxContract<mcms.ContractClient>
+  }
+
+  const MCMS_NUM_GROUPS = 32
+  const NUM_SIGNERS = 20
+
+  let testSigners: TestSigner[]
+
+  beforeAll(async () => {
+    code = await MCMSBaseTestSetup.compileContracts()
+  })
+
+  beforeEach(async () => {
+    blockchain = await Blockchain.create()
+    blockchain.now = 1
+
+    // Set up accounts
+    acc = {
+      deployer: await blockchain.treasury('deployer'),
+      multisigOwner: await blockchain.treasury('multisigOwner'),
+      signers: [],
+    }
+
+    // Generate deterministic test signers
+    testSigners = []
+    {
+      let keyPairs = await Promise.all(
+        Array.from({ length: NUM_SIGNERS }, async (_, i) => await generateEd25519KeyPair()),
+      )
+
+      // Sort result by public key (strictly increasing)
+      keyPairs.sort((a, b) => {
+        const aKey = uint8ArrayToBigInt(a.publicKey)
+        const bKey = uint8ArrayToBigInt(b.publicKey)
+        return aKey < bKey ? -1 : aKey > bKey ? 1 : 0
+      })
+
+      for (let i = 0; i < NUM_SIGNERS; i++) {
+        const treasury = await blockchain.treasury(`signer${i}`)
+        // This is a simplified approach - in real tests you might want to use actual key generation
+        const address = treasury.address
+
+        testSigners.push({
+          address,
+          keyPair: keyPairs[i],
+          treasury: treasury,
```

Comment:
nit: we could call this `wallet` which is more common, and drop the address field